test/e2e/packageinstall_test.go

```diff
@@ -338,3 +338,108 @@ spec:
 		kubectl.Run([]string{"delete", "pkgi", name, "--wait=true"})
 	})
 }
+
+func Test_PackageInstall_UsesExistingAppWithSameName(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, logger}
+	kubectl := Kubectl{t, env.Namespace, logger}
+	sas := ServiceAccounts{env.Namespace}
+	name := "pkg-instl-uses-app"
+
+	appYaml := fmt.Sprintf(`
+---
+apiVersion: kappctrl.k14s.io/v1alpha1
+kind: App
+metadata:
+  name: %s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/apps
+spec:
+  serviceAccountName: kappctrl-e2e-ns-sa
+  fetch:
+    - inline:
+        paths:
+          file.yml: |
+            apiVersion: v1
+            kind: ConfigMap
+            metadata:
+              name: configmap
+  template:
+  - ytt: {}
+  deploy:
+    - kapp: {}
+`, name) + sas.ForNamespaceYAML()
+
+	pkginstallYaml := fmt.Sprintf(`---
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: PackageMetadata
+metadata:
+  name: pkg.test.carvel.dev
+  namespace: %[1]s
+spec:
+  displayName: "Test PackageMetadata"
+  shortDescription: "PackageMetadata used for testing"
+---
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  name: pkg.test.carvel.dev.1.0.0
+  namespace: %[1]s
+spec:
+  refName: pkg.test.carvel.dev
+  version: 1.0.0
+  template:
+    spec:
+      fetch:
+      - inline:
+          paths:
+            file.yml: |
+              apiVersion: v1
+              kind: ConfigMap
+              metadata:
+                name: configmap
+      template:
+      - ytt: {}
+      deploy:
+      - kapp: {}
+---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: %[2]s
+  namespace: %[1]s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/packageinstalls
+spec:
+  serviceAccountName: kappctrl-e2e-ns-sa
+  packageRef:
+    refName: pkg.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+`, env.Namespace, name) + sas.ForNamespaceYAML()
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+	cleanUp()
+	defer cleanUp()
+
+	// Create App
+	kubectl.RunWithOpts([]string{"apply", "-f", "-"}, RunOpts{StdinReader: strings.NewReader(appYaml)})
+
+	// Create PackageInstall
+	kapp.RunWithOpts([]string{"deploy", "-a", name, "-f", "-"}, RunOpts{StdinReader: strings.NewReader(pkginstallYaml)})
+	kubectl.Run([]string{"wait", "--for=condition=ReconcileSucceeded", "pkgi/" + name, "--timeout", "1m"})
+
```

Comment:
i think we need stronger assertion that App CR is "taken over". how about making Package CR contain some different in spec.template and checking that App CR has that change after PackageInstall is reconciled?

Reply:
Sure, I am going to recommend keeping #184 open after feedback is implemented though as I still think this behavior of PackageInstalls taking over Apps is not intuitive as far as UX is concerned. But I agree that the assertion can be made stronger.